Refactor logout

code-asher opened the pull request:
This should fix logging out on non-localhost domains and non-root paths (and some other cases like when using `--proxy-domain`).

It also moves the logout to a command and registers it from our client entry so we don't need to patch VS Code. This new command redirects to the logout endpoint where we clear the cookie from the server end.

Fixes #2984

Should unblock https://github.com/cdr/code-server/discussions/2245#discussioncomment-614259 as well

code-asher commented:
Moving to a draft while I update the tests.

jsjoeio commented on typings/ipc.d.ts:
```diff
@@ -6,9 +6,12 @@
  * The second is a symlink to the first.
  */
 export interface Options {
+  authed: boolean
   base: string
+  csStaticBase: string
   disableTelemetry: boolean
   disableUpdateCheck: boolean
+  logLevel: number
```
This may break some tests in `util.test.ts` 🤔 https://github.com/cdr/code-server/blob/main/test/unit/util.test.ts#L127-L191

code-asher replied:
These properties were actually always there, just not in the types, so we might be OK.

code-asher commented on lib/vscode/src/vs/server/browser/client.ts:
```diff
@@ -173,38 +163,36 @@ export const initialize = async (services: ServiceCollection): Promise<void> =>
+	CommandsRegistry.registerCommand(
+		LOGOUT_COMMAND_ID,
+		() => {
+			window.location.href = logoutEndpoint;
```
I tried actually but the redirect seems to happens before the log can be sent over the socket. :frowning: 

We could do a `console.log` but that'll get lost on a page reload unless you have persistent logging enabled.

Maybe what we should do is pop up a notification asking if you really want to log out? And show the path it'll open as well.

jsjoeio replied:
Ah bummer! No worries at all. 

> Maybe what we should do is pop up a notification asking if you really want to log out? And show the path it'll open as well.

Totally your call. We'd have to adjust the e2e tests if we did that.

code-asher replied:
Hmmm...I think we leave it for a future enhancement for now.

jsjoeio commented:
> will increase coverage by 7.87%.

DANG! Look at you 👏

jsjoeio reviewed (approved):
![You did a good job. False. You did an awesome job.](https://user-images.githubusercontent.com/3806031/117057265-17f56580-acd2-11eb-9dfc-950ba67cfc4b.png)

oxy commented:
The one thing I'm slightly concerned about is that we have a 1.56 branch in the pipeline already which makes its own set of changes to menubarControl.ts, etc. - how much of this will we have to reimplement/re-fix in the 1.56 branch?

code-asher commented:
Ah yup actually I initially based this off your branch and switched back to main so I could get it merged in. :stuck_out_tongue:

I removed all the changes to that file so all I did was checkout the source. We can do that again in the 1.56 branch. Or we could resolve manually, I think the only conflict was the `logService` line and we'd want to remove the comment added above as well.
